### PR TITLE
Dynamic service switching

### PR DESCRIPTION
**Pull request by martin-schaefer**
Hi

I've implemented the dynamic service switching using a ServiceTracker in DynamicInvocationHandler. The itest covering the new behaviour is working but I have two concerns:

- The SerivceTracker is opened in DynamicInvocationHandler(). I tried to close it in DynamicInvocationHandler.release() but this seems to be too early and the test fails. Is there a correct place to close it?
- Since in DynamicInvocationHandler the ServiceObjectsWrapper is not used anymore, the behaviour for OSGi 6 is probably not correct. The ServiceTracker should probably move into the Osgi5ServiceObjectsWrapper but what has to be done in the Osgi6ServiceObjectsWrapper? I am not familiar with service scopes in OSGi 6.

Regards,
Martin

### CODE REVIEW

File: pax-cdi-extension/src/main/java/org/ops4j/pax/cdi/extension/impl/client/AbstractServiceInvocationHandler.java [hunk 1]
```diff
@@ -12,29 +12,26 @@
...
-        ServiceReference<CdiContainerFactory> serviceReference = bundleContext
-            .getServiceReference(CdiContainerFactory.class);
-        CdiContainerFactory cdiContainerFactory = bundleContext.getService(serviceReference);
-        this.cdiContainer = cdiContainerFactory.getContainer(bundleContext.getBundle());
-    }
-
-    public abstract void release();
-
+	
+	protected InjectionPoint			ip;
```
Comment (ANierbeck) [inline pax-cdi-extension/src/main/java/org/ops4j/pax/cdi/extension/impl/client/AbstractServiceInvocationHandler.java]: just a minor comment, those extra white spaces here don't help much ;)

Comment (martin-schaefer) [inline pax-cdi-extension/src/main/java/org/ops4j/pax/cdi/extension/impl/client/AbstractServiceInvocationHandler.java]: Sorry for the whitespaces. I had my Eclipse autoformat turned on. Are these [coding styles](https://ops4j1.jira.com/wiki/display/ops4j/Coding+Style) still valid?

Comment (ANierbeck) [inline pax-cdi-extension/src/main/java/org/ops4j/pax/cdi/extension/impl/client/AbstractServiceInvocationHandler.java]: no problem ;)
regarding the coding styles I'm not sure they are a bit outdated. 
but there are some checkstyle rules (though I don't know if they check this :-) ) 
https://github.com/ops4j/org.ops4j.pax.cdi/tree/master/pax-cdi-checkstyle-rules

**Comment by hwellmann** [comment]
Pax CDI uses the Eclipse formatter of Pax Exam:

https://github.com/ops4j/org.ops4j.pax.exam2/blob/master/assets/EclipseJavaFormatter.xml

**Comment by martin-schaefer** [comment]
Any chance someone will review this pull request?

**Comment by hwellmann** [comment]
Please be patient, we're all doing this in our spare time...

**Comment by hwellmann** [comment]
Fixed by https://ops4j1.jira.com/browse/PAXCDI-166, so closing this pull request.

As pointed out in previous comments, this PR violated our coding styles (indentation, spurious `this` and `final`) and it broke some tests for OSGi 6 prototype scope.

I created a patch from this and modified it to make things work.

So thanks again for the contribution! And next time please mind the proverb _When in Rome, do as Romans do_ - you don't have to like our style, but we like to keep it consistent.

**Pull request closed without merging**